src/dos/dos_memory.cpp

```diff
@@ -504,33 +504,45 @@ void DOS_SetupMemory(void) {
 	mcb_sizes+=17;
 	tempmcb2.SetType(middle_mcb_type);
 
-	DOS_MCB mcb((uint16_t)DOS_MEM_START+mcb_sizes);
-	mcb.SetPSPSeg(MCB_FREE);						//Free
-	mcb.SetType(ending_mcb_type); // Last Block
 	if (machine==MCH_TANDY) {
 		/* memory up to 608k available, the rest (to 640k) is used by
 			the tandy graphics system's variable mapping of 0xb800 */
-		mcb.SetSize(0x9BFF - DOS_MEM_START - mcb_sizes);
+		DOS_MCB free_block((uint16_t)DOS_MEM_START+mcb_sizes);
+		free_block.SetPSPSeg(MCB_FREE);
+		free_block.SetType(ending_mcb_type);
+		free_block.SetSize(0x9BFF - DOS_MEM_START - mcb_sizes);
 	} else if (machine==MCH_PCJR) {
-		/* memory from 128k to 640k is available */
-		mcb_devicedummy.SetPt((uint16_t)0x2000);
-		mcb_devicedummy.SetPSPSeg(MCB_FREE);
-		mcb_devicedummy.SetSize(umb_start_seg - 0x2000);
-		mcb_devicedummy.SetType(ending_mcb_type);
-
-		/* exclude PCJr graphics region */
-		mcb_devicedummy.SetPt((uint16_t)0x17ff);
-		mcb_devicedummy.SetPSPSeg(MCB_DOS);
-		mcb_devicedummy.SetSize(0x800);
-		mcb_devicedummy.SetType(middle_mcb_type);
-
-		/* memory below 96k */
-		mcb.SetSize(0x1800 - DOS_MEM_START - (2+mcb_sizes));
-		mcb.SetType(middle_mcb_type);
+		// PCjr video memory uses shared RAM at linear address 96KB - 128KB
+		// This video memory is mapped in 16KB pages
+		// Sierra AGI games crash if the application memory is not contiguous
+		// Work around that by reserving everything up until the application segment
+		const uint16_t reserved_segment = DOS_MEM_START + mcb_sizes;
+
+		// Segment at which application memory starts.
+		// (0x2400 * 16 bytes per segment) / 1024 bytes per kilobyte = 144KB linear address
+		// Space Quest version 1.0x is a special case that requires an additional 16 KB page to be reserved
+		const uint16_t application_segment = 0x2400;
```

Comment:
How about something like this to turn the comment into actual code?

```cpp
constexpr auto bytes_per_segment = 16;
constexpr auto application_segment = 144 /* kilobytes */ * 1024 / bytes_per_segment
```

Reply:
Does this mean the video mem is usually 32KB in total on the PCjr, from 96 to 128 KB, but because SQ v1.0 requires an extra 16 KB of video memory, that pushed our application segment start offset from 144KB from 128KB? I had to guess this, so I think the comments should be a bit clearer.

Reply:
Yep, that's basically the situation.  SQ 1.0x is a strange case because the video memory of PCjr cannot extend past 128KB.  As far as I can tell, that's a hardware limitation and the code in `vga_other.cpp` seems to replicate that.

I'm not sure what exactly it's doing with that extra 16KB.  Maybe using it for a backbuffer or something.  I just kept running into what looked like memory corruption bugs until I raised it.  The video output looked fine but there were crashes and disk errors and such.

I'll clarify like you suggested.